bin/src/learner/raytune_learner.py

```diff
@@ -0,0 +1,243 @@
+        """
+        Run the tuning process.
+        """
+        if overwrite:
+            self.tuner = None
+            self.best_config = None
+            
+        if self.tuner is None:
+            self._prep_tuner()
+            self.results = self.tuner.fit()
+            best_config = os.path.join(self.results.get_best_result().path, "params.json")
+            try: 
+                assert os.path.exists(best_config)
+            except AssertionError:
+                raise ValueError("Best config file not found.")
+            with open(best_config, "r") as f:
+                self.best_config = eval(f.read())
+        else:
+            raise ValueError("Tuner already exists - if you want to overwrite it, please set overwrite=True.")     
+
+    def store_best_config(self, path):
+        """
+        Store the best config in a file.
+        """
+        with open(path, "w") as f:
+            f.write(str(self.best_config))            
+    
+    def train(self, config = None): 
+        """
+        Train the model with the config.
+        """
+        if config is None:
+            config = self.best_config
+        try: 
+            assert config is not None
+        except AssertionError:
+            raise ValueError("No config provided - please provide a config to train the model with or tune first by calling the tune() method.")
+        
+        self.trainer = TuneModel(config=config)
+        self.trainer.tuning = False
+        for i in range(config["epochs"]):
+            self.trainer.step()
+
+
+class TuneModel(Trainable):
+
+    def setup(self, config: dict):
+        """
+        Get the model, loss function(s), optimizer, train and test data from the config.
+        """
+
+        # Load model from string path
+        module_name, class_name = config["model"].rsplit('.', 1)
+        module = importlib.import_module(module_name)
+        model = getattr(module, class_name)
+        self.model = model(**config["model_params"])
+
+        # Add data path
+        self.data_path = config["data_path"]
+
+        # Add experiment
+        module_name, class_name = config["experiment"].rsplit('.', 1)
+        module = importlib.import_module(module_name)
+        self.experiment = getattr(module, class_name)()
+
+        # Get the loss function(s) from the config model params
+        # Note that the loss function(s) are stored in a dictionary, 
+        # where the key is the name of the loss function and the value is the loss function itself.
+        self.loss_dict = config["loss_fn"]
+        for key, loss_fn in self.loss_dict.items():
+            try:
+                #self.loss_dict[key] = getattr(nn, loss_fn)()
+                self.loss_dict[key] = loss_fn
+            except AttributeError:
+                raise ValueError(f"Invalid loss function: {loss_fn}, check PyTorch for documentation on available loss functions")
+        
+        # get the optimizer from the config
+        try:
```

Comment:
can this two checks, this line and line 149 be moved in the TuneWrapper like all the other checks?

Reply:
Yes i think it could - yet i'd leave this whole moving of the checking to whoever does it for the corresponding issue so it's done once and for all. But good point!